slepc: init at 3.22.2

qbisi opened the pull request:
Scalable Library for Eigenvalue Problem Computations
Sci-lib depends on petsc and used by many math softwares.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

qbisi committed:
petsc: do not install examples by default

Petsc examples take up 58M and can be found in source code.
Thus we do not ship petsc with examples by default.

NickCao commented on pkgs/by-name/pe/petsc/package.nix:
```diff
@@ -174,10 +174,14 @@ stdenv.mkDerivation rec {
-  nativeInstallCheckInputs = [ mpiCheckPhaseHook ];
+  nativeInstallCheckInputs = [
+    mpiCheckPhaseHook
+  ] ++ lib.optional pythonSupport python3Packages.pythonImportsCheckHook;
```
Use `lib.optionals`

qbisi committed:
petsc: remove redundant configureScript

shebang in configure script is auto patched

NickCao commented on pkgs/by-name/pe/petsc/package.nix:
```diff
@@ -90,20 +91,16 @@ stdenv.mkDerivation rec {
-        --replace /usr/bin/install_name_tool ${cctools}/bin/install_name_tool
-    '';
+    substituteInPlace config/example_template.py \
+      --replace-fail "/usr/bin/env bash" "${bash}/bin/bash"
```
Can we use `patchShenbangs`?

qbisi replied:
this fix the string in example template not the shebang of the file example_template.py itself

qbisi replied:
```
head example_template.py
header="""#!/usr/bin/env bash
# This script was created by gmakegentest.py

@COMMENTS@

@ENV@
```

NickCao commented on pkgs/by-name/sl/slepc/package.nix:
```diff
@@ -0,0 +1,100 @@
+
+  nativeInstallCheckInputs = [
+    mpiCheckPhaseHook
+  ] ++ lib.optional pythonSupport python3Packages.pythonImportsCheckHook;
```
Use `lib.optionals`

NickCao commented on pkgs/by-name/sl/slepc/package.nix:
```diff
@@ -0,0 +1,100 @@
+  doInstallCheck = true;
+
+  preInstallCheck = ''
+    export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:$out/lib
```
Why is this necessary? Do the resulting binary work without LD_LIBRARY_PATH set?

qbisi replied:
I forgot to delete it. sorrry.

qbisi committed:
sowing: init at 1.1.26.12

qbisi committed:
slepc: init at 3.22.2

qbisi committed:
python312packages.petsc4py: do not build withFullDeps

The `withFullDeps` flag was used to make petsc4py meet the requirements of downstream
projects like Fenics and Firedrake. However, this flag breaks the build on other platforms
if any external packages of PETSc fail.

Therefore, I disabled this flag in petsc4py and made PETSc bundle some commonly used
external packages by default. Users can still have a minimally featured PETSc by setting
`withCommonDeps = false`.

qbisi committed:
petsc: fix and add python related tests

doronbehar commented on pkgs/by-name/pe/petsc/package.nix:
```diff
@@ -92,20 +92,16 @@ stdenv.mkDerivation rec {
-      patchShebangs ./lib/petsc/bin
-    ''
-    + lib.optionalString stdenv.hostPlatform.isDarwin ''
-      substituteInPlace config/install.py \
-        --replace /usr/bin/install_name_tool ${cctools}/bin/install_name_tool
```
> substitute install_name_tool replace nothing but echo quiet

This part of the Git commit is not clear to me. Do you mean the substitution hasn't done anything?

qbisi replied:
yes, it is an old patch for petsc, and it will raise error if you use --replace-fail.
now the install.py will just use install_name_tool in system path.
so no need to patch this.

doronbehar replied:
> yes, it is an old patch for petsc, and it will raise error if you use --replace-fail.
> now the install.py will just use install_name_tool in system path.
> so no need to patch this.

I see. You should work on your phrasings, especially if you wish they'd get into the History of Nixpkgs, and if you wish to ease the work of your reviewers. I'd also put this change in a separate commit, as it is justified and can be decoupled from other changes which are much more debatable.

qbisi replied:
I will make this a seperate commit, but not untill other debatable commits are resolved.
I found it very hard for me to rebase with so many commits.

doronbehar commented on pkgs/by-name/pe/petsc/package.nix:
```diff
@@ -92,20 +92,16 @@ stdenv.mkDerivation rec {
 
   propagatedBuildInputs = lib.optional pythonSupport python3Packages.numpy;
 
-  patches = [
-    (replaceVars ./fix-petsc4py-install-prefix.patch {
-      PYTHON_SITEPACKAGES = python3.sitePackages;
-    })
-  ];
```
I personally think there's an advantage to patches, as they are a bit more specific - for example the `'lib'` string, might be found in future releases in a different place in the file - a place you wouldn't expect and it would cause you wierd bugs that would be very hard to debug.

qbisi replied:
I will ask for more comment on this change.
Personaly, i would like to use a sinlge command substituteInplace than patch file as it keep the nixpkgs repository tidy.
I would submit patchfile to upstream if possible, but this patch is only distrution related.

doronbehar commented on pkgs/by-name/pe/petsc/package.nix:
```diff
@@ -174,10 +174,18 @@ stdenv.mkDerivation rec {
+      python3Packages.pythonImportsCheckHook
+    ];
+
+  pythonImportsCheck = [ "petsc4py" ];
```
Nit: Technically, these could be separated to a different commit, as they are not part of the argument rename.

doronbehar commented on pkgs/by-name/pe/petsc/package.nix:
```diff
@@ -174,10 +174,18 @@ stdenv.mkDerivation rec {
 
   passthru = {
-    inherit mpiSupport;
+    inherit mpiSupport pythonSupport;
```
Same nit here - could be separated to a different commit.

doronbehar replied:
> Same nit here - could be separated to a different commit.

This nit still applies :) If you won't mind grabbing that in the next iteration it'd be nice :).

doronbehar commented on pkgs/by-name/sl/slepc/package.nix:
```diff
@@ -0,0 +1,115 @@
+    hash = "sha256-a5DmsA7NAlhrEaS43TYPk7vtDfhXLEP+5sftu2A9Yt4=";
+  };
+
+  postPatch = ''
+    # fix slepc4py install prefix
+    substituteInPlace config/packages/slepc4py.py \
+      --replace-fail "slepc.prefixdir,'lib'" \
+      "slepc.prefixdir,'${python3.sitePackages}'"
+
+    patchShebangs lib/slepc/bin
+
+    # use system bfort
+    substituteInPlace config/packages/sowing.py \
+      --replace-fail "bfort = os.path.join(archdir,'bin','bfort')" \
+      "bfort = '${sowing}/bin/bfort'"
+  '';
```
Here too I'd recommend using a `replaceVars` on a patch, but it is not so bad as in my previous comment about this, because the replaced strings look much more like strings that would appear only once in the files..

doronbehar commented on pkgs/by-name/sl/slepc/package.nix:
```diff
@@ -0,0 +1,115 @@
+      "bfort = '${sowing}/bin/bfort'"
+  '';
+
+  preConfigure = ''
+    export SLEPC_DIR=$PWD
+  '';
```
Could you explain this please in a comment?

qbisi replied:
$SLEPC_DIR are common used in makefile and should be set to the source root to slepc.
It is useful if you build slepc from another working directory. e.g. configure petsc with --download-slepc=1 will download and build slepc.
When packaging slepc alone, just set it to $PWD.

doronbehar reviewed (commented):
Looks pretty good overall, nice contribution!

qbisi committed:
python312packages.petsc4py: replace withPetsc4py with pythonSupport

pythonSupport is more commonly used in toPythonModule

qbisi committed:
petsc: remove unneccessary substitue of install_name_tool

This substitueInplace was an old patch and will fail if use --replace-fail.

qbisi committed:
petsc: patch shebang in template string of example_template.py

This patch does not fix the shebang at the begining of the example_template.py file.
It only replaces the shebang in the template string of that file.

qbisi committed:
petsc: add meta.platforms

doronbehar commented on pkgs/by-name/sl/slepc/package.nix:
```diff
@@ -0,0 +1,115 @@
+    ];
+    platforms = lib.platforms.unix;
+    maintainers = with lib.maintainers; [ qbisi ];
+    broken = stdenv.hostPlatform.isDarwin && withArpack;
```
If you have it available, it'd be nice to get a small quote from the build error. Alternatively, perhaps upstream is aware of this incompatibility, and would have liked to be notified? A comment of any kind is desired here.

qbisi replied:
Segmentation error in mpi test for darwin platform, no idea how to fix this, but i will put the error message here.

doronbehar reviewed (commented):
Looks very good! Thanks for making progress. A few small issues / questions are left and this should be good to go after a nixpkgs-review report.

qbisi commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 389678`

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 13 packages built:</summary>
  <ul>
    <li>petsc</li>
    <li>pflotran</li>
    <li>precice</li>
    <li>python312Packages.petsc4py</li>
    <li>python312Packages.pyprecice</li>
    <li>python312Packages.pyprecice.dist</li>
    <li>python312Packages.slepc4py</li>
    <li>python313Packages.petsc4py</li>
    <li>python313Packages.pyprecice</li>
    <li>python313Packages.pyprecice.dist</li>
    <li>python313Packages.slepc4py</li>
    <li>slepc</li>
    <li>sowing</li>
  </ul>
</details>

qbisi commented on pkgs/by-name/sl/slepc/package.nix:
```diff
@@ -0,0 +1,115 @@
+      arpack-mpi
+    ];
+
+  propagatedBuildInputs = [
+    petsc
+  ];
```
> This is kind of surprising. I thought due to the above `SLEPC_DIR` comments that `slepc` is a dependency of `petsc`, or that neither is a dependency of the other.

This propagate requirement of petsc can be vailid in the output slepc.pc file, it also ensure petsc4py the dependecy of slepc4py when enablePython=true.

qbisi replied:
> This is kind of surprising. I thought due to the above `SLEPC_DIR` comments that `slepc` is a dependency of `petsc`, or that neither is a dependency of the other.

Slepc depend on petsc, petsc can be configured to download and install slepc as additional software after petsc is installed. Likewise, mfem, hpddm also  depend on petsc, and you can install them together with petsc by specifying --download-mfem=1 --download-hpddm=1

qbisi commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 389678`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 14 packages built:</summary>
  <ul>
    <li>getdp</li>
    <li>petsc</li>
    <li>pflotran</li>
    <li>precice</li>
    <li>python312Packages.petsc4py</li>
    <li>python312Packages.pyprecice</li>
    <li>python312Packages.pyprecice.dist</li>
    <li>python312Packages.slepc4py</li>
    <li>python313Packages.petsc4py</li>
    <li>python313Packages.pyprecice</li>
    <li>python313Packages.pyprecice.dist</li>
    <li>python313Packages.slepc4py</li>
    <li>slepc</li>
    <li>sowing</li>
  </ul>
</details>

qbisi commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 389678`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 14 packages built:</summary>
  <ul>
    <li>getdp</li>
    <li>petsc</li>
    <li>pflotran</li>
    <li>precice</li>
    <li>python312Packages.petsc4py</li>
    <li>python312Packages.pyprecice</li>
    <li>python312Packages.pyprecice.dist</li>
    <li>python312Packages.slepc4py</li>
    <li>python313Packages.petsc4py</li>
    <li>python313Packages.pyprecice</li>
    <li>python313Packages.pyprecice.dist</li>
    <li>python313Packages.slepc4py</li>
    <li>slepc</li>
    <li>sowing</li>
  </ul>
</details>

qbisi committed:
petsc: add pythonSupport to passthru attributes

qbisi commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 389678`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 14 packages built:</summary>
  <ul>
    <li>getdp</li>
    <li>petsc</li>
    <li>pflotran</li>
    <li>precice</li>
    <li>python312Packages.petsc4py</li>
    <li>python312Packages.pyprecice</li>
    <li>python312Packages.pyprecice.dist</li>
    <li>python312Packages.slepc4py</li>
    <li>python313Packages.petsc4py</li>
    <li>python313Packages.pyprecice</li>
    <li>python313Packages.pyprecice.dist</li>
    <li>python313Packages.slepc4py</li>
    <li>slepc</li>
    <li>sowing</li>
  </ul>
</details>

qbisi commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 389678`

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 13 packages built:</summary>
  <ul>
    <li>petsc</li>
    <li>pflotran</li>
    <li>precice</li>
    <li>python312Packages.petsc4py</li>
    <li>python312Packages.pyprecice</li>
    <li>python312Packages.pyprecice.dist</li>
    <li>python312Packages.slepc4py</li>
    <li>python313Packages.petsc4py</li>
    <li>python313Packages.pyprecice</li>
    <li>python313Packages.pyprecice.dist</li>
    <li>python313Packages.slepc4py</li>
    <li>slepc</li>
    <li>sowing</li>
  </ul>
</details>

doronbehar reviewed (approved):
Very well written, thanks again!

qbisi commented:
Thanks for your reivew, too.
